ibis/expr/signature.py

```diff
@@ -0,0 +1,214 @@
+          Note, that default value (except for None) must also pass the inner
+          validator.
+          If callable is passed, it will be executed just before the inner, and
+          itsreturn value will be treaded as default.
+        """
+        if PY2:
+            self._serial = next(self._counter)
+
+        self.default = default
+        if isinstance(validator, type):
+            self.validator = rlz.instance_of(validator)
+        elif isinstance(validator, tuple):
+            assert util.all_of(validator, type)
+            self.validator = rlz.instance_of(validator)
+        elif callable(validator):
+            self.validator = validator
+        else:
+            raise TypeError('Argument validator must be a callable, type or '
+                            'tuple of types, given: {}'.format(validator))
+
+    def __eq__(self, other):
+        return (
+            self.validator == other.validator and
+            self.default == other.default
+        )
+
+    def optional(self):
+        return self.default is not _undefined
+
+    def validate(self, value=_undefined, name=None):
+        """
+        Parameters
+        ----------
+        value : Any, default undefined
+          Raises TypeError if argument is mandatory but not value has been
+          given.
+        name : Optional[str]
+          Argument name for error message
+        """
+        if self.optional():
+            if value is _undefined or value is None:
+                if self.default is None:
+                    return None
+                elif util.is_function(self.default):
+                    value = self.default()
+                else:
+                    value = self.default
+        elif value is _undefined:
+            if name is not None:
+                name = ' `{}`'.format(name)
+            raise TypeError('Missing required value for argument' + name)
+
+        return self.validator(value)
+
+    __call__ = validate  # syntactic sugar
+
+
+class Return(object):
+    # TODO
+    """Acts like a method (output type)"""
+
+    def __call__(self, obj):
+        pass
+
+
+class TypeSignature(OrderedDict):
+
+    __slots__ = tuple()
+
+    def validate(self, *args, **kwargs):
+        if len(args) > len(self.keys()):
+            raise TypeError('takes {} positional arguments ut {} were '
+                            'given'.format(len(self.keys()), len(args)))
+
+        result = []
+        for i, (name, argument) in enumerate(self.items()):
+            if i < len(args):
+                if name in kwargs:
+                    raise TypeError(
+                        'Got multiple values for argument {}'.format(name)
+                    )
+                value = argument.validate(args[i], name=name)
+            elif name in kwargs:
+                value = argument.validate(kwargs[name], name=name)
+            else:
+                value = argument.validate(name=name)
```

Comment:
Do you think it might be possible to `bind` arguments using `inspect.signature` here so we don't have to recreate Python's calling convention algorithm?

Reply:
Good question. Isn't `inspect.signature` py3 only?

Reply:
Yep, but there's the [funcsigs](https://pypi.python.org/pypi/funcsigs) package for that :)

Reply:
I'm going to add `funcsigs` as a dependency for https://github.com/ibis-project/ibis/pull/1277, so don't worry about adding it here

Reply:
Converted to issue: https://github.com/ibis-project/ibis/issues/1396